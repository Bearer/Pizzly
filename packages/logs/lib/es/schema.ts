import type { opensearchtypes } from '@opensearch-project/opensearch';
import type { MessageRow } from '@nangohq/types';
import { envs } from '../env.js';

const props: Record<keyof MessageRow, opensearchtypes.MappingProperty> = {
    id: { type: 'keyword' },

    parentId: { type: 'keyword' },

    accountId: { type: 'keyword' },
    accountName: { type: 'keyword' },

    environmentId: { type: 'keyword' },
    environmentName: { type: 'keyword' },

    configId: { type: 'keyword' },
    configName: { type: 'keyword' },
    providerName: { type: 'keyword' },

    connectionId: { type: 'keyword' },
    connectionName: { type: 'keyword' },

    syncConfigId: { type: 'keyword' },
    syncConfigName: { type: 'keyword' },

    jobId: { type: 'keyword' },

    userId: { type: 'keyword' },

    operation: {
        properties: {
            type: { type: 'keyword' },
            action: { type: 'keyword' }
        }
    },

    type: { type: 'keyword' },
    title: { type: 'keyword' },
    level: { type: 'keyword' },
    state: { type: 'keyword' },
    code: { type: 'keyword' },

    source: { type: 'keyword' },

    message: { type: 'text', analyzer: 'standard', search_analyzer: 'standard' },
    meta: { type: 'object', enabled: false },
    error: { type: 'object', enabled: false },

    request: {
        properties: {
            url: { type: 'keyword' },
            method: { type: 'keyword' },
            headers: { type: 'object', enabled: false }
        }
    },

    response: {
        properties: {
            code: { type: 'integer' },
            headers: { type: 'object', enabled: false }
        }
    },

    createdAt: { type: 'date' },
    updatedAt: { type: 'date' },
    startedAt: { type: 'date' },
    endedAt: { type: 'date' }
};

export const indexMessages: opensearchtypes.IndicesCreateRequest = {
    index: envs.NANGO_LOGS_OS_INDEX ?? 'messages',
    body: {
        settings: {
            analysis: {
                analyzer: {
                    default: {
                        type: 'standard'
                    },
                    default_search: {
                        type: 'standard'
                    }
                }
            }
        },
        mappings: {
            dynamic: false,
            properties: props
        }
    }
};

export const indices: opensearchtypes.IndicesCreateRequest[] = [indexMessages];
