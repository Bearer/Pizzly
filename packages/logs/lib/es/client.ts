import { Client } from '@elastic/elasticsearch';
import { envs } from '../env.js';
import { isCli } from '../utils.js';

// TODO: remove this
export const client = isCli
    ? (null as unknown as Client)
    : new Client({
          nodes: envs.NANGO_LOGS_ES_URL,
          requestTimeout: 5000,
          maxRetries: 1,
          auth: { username: envs.NANGO_LOGS_ES_USER, password: envs.NANGO_LOGS_ES_PWD },
          tls: { rejectUnauthorized: false }
      });
