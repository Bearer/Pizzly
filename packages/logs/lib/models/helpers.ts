import type { MessageRow } from '../types/messages';
import { nanoid } from '../utils.js';

export interface FormatMessageData {
    account?: { id: number; name: string };
    user?: { id: number } | undefined;
    environment?: { id: number; name?: string } | undefined;
}

export function getFormattedMessage(data: Partial<MessageRow>, { account, user, environment }: FormatMessageData = {}): MessageRow {
    return {
        id: data.id || nanoid(),

        source: data.source || 'internal',
        level: data.level || 'info',
        operation: data.operation || null,
        type: data.type || 'log',
        message: data.message || '',
        title: data.title || null,
        code: data.code || null,
        state: data.state || 'waiting',

        accountId: account?.id || data.accountId || null,
        accountName: account?.name || data.accountName || null,

        environmentId: environment?.id || data.environmentId || null,
        environmentName: environment?.name || data.environmentName || null,

        configId: data.configId || null,
        configName: data.configName || null,

        connectionId: data.connectionId || null,
        connectionName: data.connectionName || null,

        syncId: data.syncId || null,
        syncName: data.syncName || null,

        jobId: data.jobId || null,

        userId: user?.id || data.userId || null,
        parentId: data.parentId || null,

        error: data.error || null,
        request: data.request || null,
        response: data.response || null,
        meta: data.meta || null,

        createdAt: data.createdAt || new Date().toISOString(),
        updatedAt: data.updatedAt || new Date().toISOString(),
        startedAt: data.startedAt || null,
        endedAt: data.endedAt || null
    };
}

export const syncCommandToOperation = {
    PAUSE: 'pause',
    UNPAUSE: 'unpause',
    RUN: 'run',
    RUN_FULL: 'run_full',
    CANCEL: 'cancel'
} as const;
