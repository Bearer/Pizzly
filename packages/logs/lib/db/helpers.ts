import { logger } from '@nangohq/shared';
import { db } from './client';
import { schema } from './config';
import { dirname } from '../env';
import path from 'node:path';

export async function migrate(): Promise<void> {
    logger.info('[logs] migration');
    const dir = path.join(dirname, 'logs/dist/db/migrations');
    await db.raw(`CREATE SCHEMA IF NOT EXISTS ${schema}`);

    const [_, pendingMigrations] = await db.migrate.list({ directory: dir });

    if (pendingMigrations.length === 0) {
        logger.info('[logs] nothing to do');
        return;
    }

    await db.migrate.latest({ directory: dir });
    logger.info('[logs] migrations completed.');
}
