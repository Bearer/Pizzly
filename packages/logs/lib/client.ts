import type { MessageRow, MessageRowInsert, MessageMeta, OperationRowInsert } from './types/messages.js';
import { setRunning, createMessage, setFailed, setCancelled, setTimeouted, setSuccess, update } from './models/messages.js';
import type { FormatMessageData } from './models/helpers.js';
import { getFormattedMessage } from './models/helpers.js';
import type { SetRequired } from 'type-fest';
import { errorToObject, metrics, stringifyError } from '@nangohq/utils';
import { logger } from './utils.js';
import { envs } from './env.js';

interface Options {
    dryRun?: boolean;
    logToConsole?: boolean;
}

export class LogContext {
    id: string;
    dryRun: boolean;
    logToConsole: boolean;

    constructor(data: { parentId: string }, options: Options = { dryRun: false, logToConsole: true }) {
        this.id = data.parentId;
        this.dryRun = envs.NANGO_LOGS_ENABLED === true ? options.dryRun || false : true;
        this.logToConsole = options.logToConsole ?? true;
    }

    /**
     * Add more data to the parentId
     */
    async enrichOperation(data: Partial<MessageRow>): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] enrich(${JSON.stringify(data)})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await update({ id: this.id, data });
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_enrich ${stringifyError(err)}`);
        }
    }

    /**
     * ------ Logs
     */
    async log(data: MessageRowInsert): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] log(${JSON.stringify(data)})`);
        }
        if (this.dryRun) {
            return;
        }

        const start = Date.now();
        try {
            await createMessage(getFormattedMessage({ ...data, parentId: this.id }));
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_insert_in_es: ${stringifyError(err)}`);
        } finally {
            metrics.duration(metrics.Types.LOGS_LOG, Date.now() - start);
        }
    }

    async debug(message: string, meta: MessageMeta | null = null): Promise<void> {
        await this.log({ type: 'log', level: 'debug', message, meta, source: 'internal' });
    }

    async info(message: string, meta: MessageMeta | null = null): Promise<void> {
        await this.log({ type: 'log', level: 'info', message, meta, source: 'internal' });
    }

    async warn(message: string, meta: MessageMeta | null = null): Promise<void> {
        await this.log({ type: 'log', level: 'warn', message, meta, source: 'internal' });
    }

    async error(message: string, meta: (MessageMeta & { error?: unknown }) | null = null): Promise<void> {
        const { error, ...rest } = meta || {};
        const err = error ? { name: 'Unknown Error', message: 'unknown error', ...errorToObject(error) } : null;
        await this.log({ type: 'log', level: 'error', message, error: err ? { name: err.name, message: err?.message } : null, meta: rest, source: 'internal' });
    }

    /**
     * @deprecated Only there for retro compat
     */
    async trace(message: string, meta: MessageMeta | null = null): Promise<void> {
        await this.log({ type: 'log', level: 'debug', message, meta, source: 'internal' });
    }

    async http(message: string, data: Pick<MessageRow, 'request' | 'response' | 'meta'>): Promise<void> {
        const level: MessageRow['level'] = data.response && data.response.code >= 400 ? 'error' : 'info';
        await this.log({ type: 'http', level, message, ...data, source: 'internal' });
    }

    /**
     * ------ State
     */
    async start(): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] start(${this.id})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await setRunning({ id: this.id });
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_enrich ${stringifyError(err)}`);
        }
    }

    async failed(): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] failed(${this.id})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await setFailed({ id: this.id });
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_set_failed ${stringifyError(err)}`);
        }
    }

    async success(): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] success(${this.id})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await setSuccess({ id: this.id });
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_set_success ${stringifyError(err)}`);
        }
    }

    async cancel(): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] cancel(${this.id})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await setCancelled({ id: this.id });
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_set_cancelled ${stringifyError(err)}`);
        }
    }

    async timeout(): Promise<void> {
        if (this.logToConsole) {
            logger.info(`[debug] timeout(${this.id})`);
        }
        if (this.dryRun) {
            return;
        }

        try {
            await setTimeouted({ id: this.id });
        } catch (err) {
            // TODO: reup throw
            logger.error(`failed_to_set_timeout ${stringifyError(err)}`);
        }
    }
}

export interface OperationContextData extends FormatMessageData {
    start?: boolean;
}

/**
 * Create an operation and return a context
 */
export async function getOperationContext(
    data: OperationRowInsert,
    { start, account, user, environment }: SetRequired<OperationContextData, 'account'>,
    options?: Options
): Promise<LogContext> {
    const msg = getFormattedMessage(data, { account, user, environment });
    if (typeof start === 'undefined' || start) {
        msg.startedAt = msg.startedAt ?? new Date().toISOString();
        msg.state = msg.state === 'waiting' ? 'running' : msg.state;
    }

    try {
        if (envs.NANGO_LOGS_ENABLED && !options?.dryRun) {
            await createMessage(msg);
        } else if (options?.logToConsole !== false) {
            logger.info(`[debug] operation(${JSON.stringify(msg)})`);
        }
    } catch (err) {
        // TODO: reup throw
        logger.error(`failed_to_create_operation ${stringifyError(err)}`);
    }

    return new LogContext({ parentId: msg.id }, options);
}

/**
 * Return a context without creating an operation
 */
export function getExistingOperationContext({ id }: { id: MessageRow['id'] }): LogContext {
    if (!id) {
        logger.error('getExistingOperationContext: id is empty');
    }

    return new LogContext({ parentId: id });
}
