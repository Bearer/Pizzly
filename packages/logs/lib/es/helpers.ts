import { client } from './client.js';
import { indices } from './schema.js';
import { logger } from '@nangohq/shared';

export async function migrateMapping() {
    try {
        await Promise.all(
            indices.map((index) => {
                return client.indices.putMapping({ index: index.index, properties: index.mappings!.properties }, { ignore: [404] });
            })
        );
    } catch (err) {
        logger.error(err);
        throw new Error('failed_to_init_elasticsearch');
    }
}
