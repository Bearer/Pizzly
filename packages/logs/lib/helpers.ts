import { logger } from '@nangohq/shared';
import { db } from './db/client';
import { schema } from './db/config';

/**
 * Create daily partition.
 */
export async function createPartitions(startDay: number = 0): Promise<void> {
    logger.info('[createLogsPartition] starting');

    const from = new Date();
    from.setDate(from.getDate() + startDay);
    from.setHours(0, 0, 0, 0);

    const to = new Date(from);
    to.setHours(23, 59, 59, 999);
    to.setDate(to.getDate() + 1);

    const fromStr = from.toISOString().split('T')[0];
    const toStr = to.toISOString().split('T')[0];
    const partitionOperation = `operations_${fromStr}`;
    const partitionMessage = `messages_${fromStr}`;

    logger.info(`Creating partitions: ${partitionOperation}, ${partitionMessage}`);
    await db.raw(
        `CREATE TABLE IF NOT EXISTS "${schema}"."${partitionOperation}" PARTITION OF "${schema}"."operations" FOR VALUES FROM ('${fromStr}') TO ('${toStr}')`
    );
    await db.raw(
        `CREATE TABLE IF NOT EXISTS "${schema}"."${partitionMessage}" PARTITION OF "${schema}"."messages" FOR VALUES FROM ('${fromStr}') TO ('${toStr}')`
    );

    await db.raw(
        `CREATE INDEX IF NOT EXISTS "operations_${partitionOperation}_level" ON "${partitionOperation}" USING BTREE (account_id, environment_id, created_at DESC, level);`
    );
    await db.raw(
        `CREATE INDEX IF NOT EXISTS "operations_${partitionOperation}_state" ON "${partitionOperation}" USING BTREE (account_id, environment_id, created_at DESC, state);`
    );

    logger.info('[createLogsPartition] done');
}
