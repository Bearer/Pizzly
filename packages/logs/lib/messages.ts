import { db } from './db/client';
import type { MessageRow } from './types/messages';
import type { OperationRow } from './types/operations';

export async function listMessages(opts: { operationId: OperationRow['id']; limit: number }): Promise<MessageRow[]> {
    return await db
        .table<MessageRow>('messages')
        .select<MessageRow[]>('*')
        .where({ operation_id: opts.operationId })
        .limit(opts.limit)
        .orderBy('created_at', 'ASC');
}
