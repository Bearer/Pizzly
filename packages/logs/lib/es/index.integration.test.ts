import { beforeAll, describe, expect, it } from 'vitest';
import { migrateMapping } from './helpers.js';
import { client } from './client.js';
import { indexMessages } from './schema.js';
import { nanoid } from '@nangohq/utils';
import { logContextGetter } from '../models/logContextGetter.js';
import { getOperation } from '../models/messages.js';

// This file is sequential
describe('mapping', () => {
    const today = new Date().toISOString().split('T')[0];
    let fullIndexName: string;
    beforeAll(async () => {
        indexMessages.index = `messages-${nanoid()}`.toLocaleLowerCase();
        fullIndexName = `${indexMessages.index}.${today}`;
    });

    it('should not have an index before migration', async () => {
        await expect(
            client.indices.getMapping({
                index: fullIndexName
            })
        ).rejects.toThrow();
    });

    it('should migrate', async () => {
        await migrateMapping();
    });

    it('should not have create any index', async () => {
        await expect(
            client.indices.getMapping({
                index: indexMessages.index
            })
        ).rejects.toThrow();
    });

    it('should not have create any index', async () => {
        await expect(
            client.indices.getMapping({
                index: indexMessages.index
            })
        ).rejects.toThrow();

        await expect(
            client.indices.getMapping({
                index: fullIndexName
            })
        ).rejects.toThrow();
    });

    it('should create one index automatically on log', async () => {
        // Log to automatically create an index
        await logContextGetter.create({ message: 'hello', operation: { type: 'action' } }, { account: { id: 1, name: '' }, environment: { id: 1, name: '' } });

        // Should have created a today index
        const mapping = await client.indices.getMapping({
            index: fullIndexName
        });
        expect(mapping[fullIndexName]).toMatchSnapshot();

        const settings = await client.indices.getSettings({
            index: fullIndexName
        });
        expect(settings[fullIndexName]?.settings?.index?.analysis).toMatchSnapshot();
        expect(settings[fullIndexName]?.settings?.index?.sort).toMatchSnapshot();
        expect(settings[fullIndexName]?.settings?.index?.lifecycle).toMatchSnapshot();
    });

    it('should create yesterday index automatically', async () => {
        const yesterday = new Date();
        yesterday.setDate(yesterday.getDate() - 1);
        const yesterdayIndexName = `${indexMessages.index}.${yesterday.toISOString().split('T')[0]}`;

        // Log to automatically create an index
        const ctx = await logContextGetter.create(
            { message: 'hello', operation: { type: 'action' }, createdAt: yesterday.toISOString() },
            { account: { id: 1, name: '' }, environment: { id: 1, name: '' } }
        );
        await ctx.failed();

        // Should have created a today index
        await client.indices.getMapping({
            index: yesterdayIndexName
        });
        const doc = await getOperation({ id: ctx.id });
        expect(doc.state).toBe('failed');
    });
});
