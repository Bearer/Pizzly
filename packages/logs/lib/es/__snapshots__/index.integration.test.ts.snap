// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mapping > should apply mapping 1`] = `
{
  "dynamic": "false",
  "properties": {
    "accountId": {
      "type": "keyword",
    },
    "accountName": {
      "type": "keyword",
    },
    "code": {
      "type": "keyword",
    },
    "integrationId": {
      "type": "keyword",
    },
    "integrationName": {
      "type": "keyword",
    },
    "connectionId": {
      "type": "keyword",
    },
    "connectionName": {
      "type": "keyword",
    },
    "createdAt": {
      "type": "date",
    },
    "endedAt": {
      "type": "date",
    },
    "environmentId": {
      "type": "keyword",
    },
    "environmentName": {
      "type": "keyword",
    },
    "error": {
      "enabled": false,
      "type": "object",
    },
    "id": {
      "type": "keyword",
    },
    "jobId": {
      "type": "keyword",
    },
    "level": {
      "type": "keyword",
    },
    "message": {
      "analyzer": "standard",
      "type": "text",
    },
    "meta": {
      "enabled": false,
      "type": "object",
    },
    "operation": {
      "properties": {
        "action": {
          "type": "keyword",
        },
        "type": {
          "type": "keyword",
        },
      },
    },
    "parentId": {
      "type": "keyword",
    },
    "providerName": {
      "type": "keyword",
    },
    "request": {
      "properties": {
        "headers": {
          "enabled": false,
          "type": "object",
        },
        "method": {
          "type": "keyword",
        },
        "url": {
          "type": "keyword",
        },
      },
    },
    "response": {
      "properties": {
        "code": {
          "type": "integer",
        },
        "headers": {
          "enabled": false,
          "type": "object",
        },
      },
    },
    "source": {
      "type": "keyword",
    },
    "startedAt": {
      "type": "date",
    },
    "state": {
      "type": "keyword",
    },
    "syncConfigId": {
      "type": "keyword",
    },
    "syncConfigName": {
      "type": "keyword",
    },
    "title": {
      "type": "keyword",
    },
    "type": {
      "type": "keyword",
    },
    "updatedAt": {
      "type": "date",
    },
    "userId": {
      "type": "keyword",
    },
  },
}
`;
